lib/watch.js

```diff
@@ -381,7 +397,14 @@ Watcher.prototype.watchDirectory = function(dir, options, fn, counter = nullCoun
 
     if (options.recursive && !has) {
       getSubDirectories(dir, function(d) {
-        self.watchDirectory(d, options, null, counter);
+        if (is.func(options.filter)) {
+          // Check to see whether to skip this directory.
+          if (options.filter(d, SKIP_FLAG) !== SKIP_FLAG) {
```

Comment:
This could be simplified to:

```js
  if (!is.func(options.filter) || options.filter(d, SKIP_FLAG) !== SKIP_FLAG) {
    self.watchDirectory(d, options, null, counter);
  }
```

I generally dislike "clever" tricks that make code harder to read, so if this smells like that to you then let's not, I don't mind.

For me, I would actually prefer this shorter way as I think it makes it clearer that there is only one decision to be made here, and that the alternate outcomes are the same.

Reply:
No, it's more clearer in that way and it removes the repeated `self.watchDirectory` 👍 . Should be fixed now!